File: src/server/src/test/test_velvet_dawn/test_mechanics/test_function_value.py
import velvet_dawn
from test.base_test import BaseTest
from velvet_dawn.dao import app
from velvet_dawn.mechanics.function_value import FunctionValue
from velvet_dawn.mechanics.selectors import Selector


"""
Test the FunctionValue works across raw values, selectors
and random values.
"""


class TestFunctionValue(BaseTest):

    def test_function_value(self):
        with app.app_context():
            self.prepare_game()

            self.assertEqual(4, FunctionValue().parse("0", 4).value(None))
            self.assertEqual("test", FunctionValue().parse("0", "test").value(None))
            self.assertEqual(None, FunctionValue().parse("0", None).value(None))
            self.assertTrue(1 >= FunctionValue().parse("0", "__rand__").value(None) >= 0)
            self.assertNotEqual(
                FunctionValue().parse("0", "__rand__").value(None),
                FunctionValue().parse("0", "__rand__").value(None)
            )

            unit = velvet_dawn.units.list()[0]
            function = FunctionValue().parse("0", "@self")
            self.assertIsInstance(function._selector_value, Selector)
            self.assertEqual(unit.id, function.value(unit))

            unit.set_attribute("example", 5)
            self.assertEqual(5, FunctionValue().parse("0", "@self.example").value(unit))

File: src/jserver/src/test/velvetdawn/mechanics/TestFunctionValue.java
package test.velvetdawn.mechanics;

import static org.junit.Assert.*;

import org.junit.Test;
import test.BaseTest;
import velvetdawn.mechanics.FunctionValue;
import velvetdawn.mechanics.selectors.Selector;
import velvetdawn.models.anytype.Any;

import java.util.ArrayList;


public class TestFunctionValue extends BaseTest {

    /*
    Test the FunctionValue works across raw values, selectors
    and random values.
    */

    @Test
    public void test_function_value() throws Exception {
        var velvetDawn = this.prepareGame();

        assertEquals(4, new FunctionValue(velvetDawn, "", Any.from(4))
                .value(null).toNumber(), 0);

        assertEquals("test", new FunctionValue(velvetDawn, "", Any.from("test"))
                .value(null).toString());

        assertNull(new FunctionValue(velvetDawn, "0", Any.Null()).value(null).toString());
        assertTrue(1 >= new FunctionValue(velvetDawn, "0", Any.from("__rand__")).value(null).toNumber());
        assertNotEquals(
                new FunctionValue(velvetDawn, "0", Any.from("__rand__")).value(null).toNumber(),
                new FunctionValue(velvetDawn, "0", Any.from("__rand__")).value(null).toNumber()
        );

        var entity = velvetDawn.entities.list().get(0);
        var function = new FunctionValue(velvetDawn, "0", Any.from("@self"));
        assertTrue(function.selectorValue instanceof Selector);
        assertEquals(entity.datapackId, function.value(entity).toString());

        entity.attributes.set("example", Any.from(5));
        assertEquals(5, new FunctionValue(velvetDawn, "0", Any.from("@self.example")).value(entity).toNumber(), 0);
    }
}
